src/i18n.ts

```diff
@@ -0,0 +1,316 @@
+      source?: string;
+      isDefault?: boolean;
+      bundleOptions?: FluentBundleOptions;
+    },
+  ): void {
+    this.fluent.addTranslationSync({
+      locales: locale,
+      isDefault: locale === this.config.defaultLocale,
+      bundleOptions: this.config.fluentBundleOptions,
+      ...options,
+    });
+
+    this.locales.push(locale);
+  }
+
+  /**
+   * Gets a message by its key from the specified locale.
+   * Alias of `translate` method.
+   */
+  t(
+    key: string,
+    locale: LocaleId,
+    context?: TranslationContext,
+  ): string {
+    return this.translate(key, locale, context);
+  }
+
+  /** Gets a message by its key from the specified locale. */
+  translate(
+    key: string,
+    locale: LocaleId,
+    context?: TranslationContext,
+  ): string {
+    return this.fluent.translate(locale, key, context);
+  }
+
+  /** Returns a middleware to .use on the `Bot` instance. */
+  middleware(): MiddlewareFn<C & I18nFlavor> {
+    return middleware(this.fluent, this.config);
+  }
+}
+
+function middleware<C extends Context = Context>(
+  fluent: Fluent,
+  { defaultLocale, localeNegotiator, useSession }: I18nConfig<C>,
+) {
+  return async function (
+    ctx: C & I18nFlavor,
+    next: NextFunction,
+  ): Promise<void> {
+    let translate: TranslateFunction;
+
+    function useLocale(locale: LocaleId): void {
+      translate = fluent.withLocale(locale);
+    }
+
+    async function getNegotiatedLocale(): Promise<string> {
+      return await localeNegotiator?.(ctx) ??
+        // deno-lint-ignore no-explicit-any
+        (await (useSession && (ctx as any).session))?.__language_code ??
+        ctx.from?.language_code ??
+        defaultLocale;
+    }
+
+    async function setLocale(locale: LocaleId): Promise<void> {
+      if (!useSession) {
+        throw new Error(
+"You are calling `ctx.i18n.setLocale()` without setting `useSession` to `true` \
+in the configuration. It doesn't make sense because you cannot set a locale in \
+the session that way. When you call `ctx.i18n.setLocale()`, the bot tries to \
+store the user locale in the session storage. But since you don't have session \
+enabled, it cannot store the locale information in the session storage. You \
+should either enable sessions or use `ctx.i18n.useLocale()` instead.",
+        );
+      }
+
+      // deno-lint-ignore no-explicit-any
+      (await (ctx as any).session).__language_code = locale;
+      await negotiateLocale();
+    }
+
+    // Determining the locale to use for translations
+    async function negotiateLocale(): Promise<void> {
+      const negotiatedLocale = await getNegotiatedLocale();
+      useLocale(negotiatedLocale);
+    }
+
+    // Also exports ctx object properties for accessing them directly from
+    // the translation source files.
+    function translateWrapper(
+      key: string,
+      context?: TranslationContext,
+    ): string {
+      return translate(key, {
+        first_name: ctx.from?.first_name ?? "",
```

Comment:
why so specific? 😄

Reply:
If I recall correctly, the `first_name` is used frequently so we go with it. I actually use it a lot as well.

Reply:
I'd rather add the full name shortcut (first name and last name if user has one or just last name otherwise) or don't add anything

Reply:
Well, we can add `last_name` as well but `full_name` is kind of overkill since that is not something we use to address each other in daily talk.

Reply:
Adding last_name is not useful at all because you can't just write `{{$first_name}} {{$last_name}} something`. last name can be absent and it would result in a double space.
My approach is not to provide any shortcuts at all and let users do it themselves.

Reply:
Correct me if I'm wrong, in some culture, address first name without being close can be considered as impolite. That is why I agree for last name but not full name since you can do more with 2 than 1 big text.

Reply:
Maybe we should remove this I think. If they need shortcuts, `CTX("from.first_name")` exists.

Reply:
Yes, I think this was a bad design to just only including the first_name. And adding a lot more would make it bloated. I'm gonna remove it.

Reply:
@dcdunkan how about we add a global variable option so dev can add set whatever they want? For example:

```Typescript
const i18n = new I18n({
	directory: './src/l10n',
	globalVars: (ctx) => ({
		first_name: ctx.from?.first_name
	}),
})
```